scripts/bootstrap.py

```diff
@@ -0,0 +1,355 @@
+        tenant_id = oidc.get("tokenUrl").split("/")[-3]
+        print(f"  [bold green]OK[/]: Azure Entra (aka ActiveDirectory) with tenant id ({tenant_id}).")
+    elif "auth0.com" in oidc.get("tokenUrl"):
+        tenant_id = oidc.get("tokenUrl").split("/")[2].split(".")[0]
+        print(f"  [bold green]OK[/] - Auth0 with tenant id ({tenant_id}).")
+    else:
+        print(f"  [bold yellow]WARNING[/]: Unknown identity provider {oidc.get('tokenUrl')}")
+    accessClaims = [c.get("claimName") for c in oidc.get("accessClaims", {})]
+    print(
+        f"  Matching on CDF group sourceIds will be done on any of these claims from the identity provider: {accessClaims}"
+    )
+    print("Checking CDF groups...")
+    try:
+        groups = ToolGlobals.client.iam.groups.list().data
+    except Exception:
+        print("  [bold red]ERROR[/]: Unable to retrieve CDF groups.")
+        ToolGlobals.failed = True
+        return
+    print(
+        Panel(
+            Columns(
+                [f"{g.id} {g.name} {g.source_id}" for g in groups],
+                title="CDF Group ids, Names, and Source Ids",
+            )
+        )
+    )
+    if len(groups) > 1:
+        print(
+            "  [bold yellow]WARNING[/]: This service principal/application gets its access rights from more than one CDF group."
+        )
+        print("          This is not recommended.")
+        if group_id == 0 and update_group:
+            print(
+                "  [bold red]ERROR[/]: You have specified --update-group. "
+                + "         With multiple groups available, you must use the --group-id option to specify which group to update."
+            )
+            ToolGlobals.failed = True
+            return
+    else:
+        print("  [bold green]OK[/] - Only one group is used for this service principal/application.")
+    print("---------------------")
+    print(f"\nChecking CDF groups access right against capabilities in {group_file} ...")
+    read_write = Group.load(
+        yaml.load(
+            Path(f"{Path(__file__).parent.parent.as_posix()}{group_file}").read_text(),
+            Loader=yaml.Loader,
+        )
+    )
+    error = False
+    all_acls = []
+    for g in groups:
+        all_acls.extend(g.capabilities)
+    for cap in read_write.capabilities:
+        all_ok = False
+        if type(cap) not in [type(a) for a in all_acls]:
+            print(
+                f"  [bold yellow]WARNING[/]: The capability {cap._capability_name} is not present in the CDF project."
+            )
+            error = True
+            continue
+        for a in all_acls:
+            all_ok = True
+            if type(a) is not type(cap):
+                continue
+            for action in cap.actions:
+                if action in a.actions:
+                    continue
+                all_ok = False
+                break
+            if all_ok:
+                break
+        if not all_ok:
```

Comment:
Fyi, this will soon become very much simpler. This might also fail wrongly when more than one group has the same type of capability.

Reply:
Actually, even with just one group you might have several different definitions of the same capability.

Reply:
Yes, the way this is built up is cumbersome... It's always the challenge of dataset and space scope vs all as well.